File: dna.py
import random
import string
from copy import deepcopy

""" 

// Genetic Algorithm, Evolving Shakespeare

// A class to describe a psuedo-DNA, i.e. genotype
//   Here, a virtual organism's DNA is an array of character.
//   Functionality:
//      -- convert DNA into a string
//      -- calculate DNA's "fitness"
//      -- mate DNA with another set of DNA
//      -- mutate DNA


"""


class DNA() :
    
    def __init__(self, genes=[""], fitness_score=0, dna_size=150) -> None:
        
        
        self.fitness_score = fitness_score 
        self.dna_size = dna_size
        self.genes = genes*self.dna_size
        self.alphabet =  list(chr(random.randint(32,127)))
        self.scoring = 0
        #list(string.ascii_lowercase)  + [" "] + ["."] + list(string.ascii_uppercase)
        
    #1- generate random dna       
    def dna(self) :
        for i in range(self.dna_size) :
            self.genes[i] = random.choices(self.alphabet,k=1)[0] 
        
    #2- Fitness function (returns floating point % of "correct" characters)
    def fitness(self, target) :
        self.score = 0
        assert(self.dna_size == len(target) ) 
        for i in range(self.dna_size) :
            if self.genes[i] == target[i] :
                self.score += 1
        
        self.scoring = self.score
        self.fitness_score = self.score / len(target)
        
    #3 crossover
    def crossover(self, dna_partner) :
        
        #3.1 create a child
        child = DNA(dna_size=len(self.genes))
        
        #3.2 pick mid point randomly
        midpoint =  random.choices([i for i in range(len(self.genes))])[0]
        
        #3.3 child is the half from both parent from the midpoint
        for i in range(len(self.genes)) :
            if i > midpoint :
                child.genes[i] = self.genes[i]
            else :
                child.genes[i] = dna_partner.genes[i] 
                
        return child
    
    
    #4 mutate the gene based on mutation rate
    def mutation(self, mutationRate,target) :

        for i in range(len(self.genes)) :
            
            if random.random() < mutationRate : #and self.genes[i] != target[i] :
                self.genes[i] = random.choices(self.alphabet,k=1)[0] 
        
         
    # 5 - get the phrase  
    def getPhrase(self) -> string :
        return ''.join(self.genes)

File: DNA.cpp
#include <iostream>
#include "DNA.hpp"
#include <random>


DNA::DNA()
{

}

DNA::DNA(int dna_size)
{

    setDNA(dna_size);

     
    for(int i =0; i < gene_size; i++){
        genes.push_back(random(32, 127));
    }
    

}
DNA::~DNA()
{

}

void DNA:: setDNA(int num) {
      gene_size = num;
    }

std::string  DNA:: GENES () {

    return genes ;
}

int DNA::getDNA() {
      return gene_size;
    }

void DNA::fitness(std::string target){
    float score = 0.0;
    for(int i= 0; i < target.length(); i++){
        if (target[i] == genes[i]){
            score++;
           
        }

    }
    
    fitness_score = score / ((float)target.length()) ;
    

}


DNA* DNA::crossover( DNA& partner){

int midpoint = random(gene_size-1) ;
DNA* child = new DNA(gene_size) ;

for (int i = 0; i < gene_size ; i++) {
    //std::cout << "\nddddbm :"<<genes[i] << " :"<< gene_size <<" :" << partner.genes[i]<<std::endl;
    if (i > midpoint) {
        child->genes[i] = genes[i] ;
    }
    child->genes[i] = partner.genes[i] ;

    

}

return child;

}

void DNA::mutate(float mutationRate) {
    int genes_size = getDNA() ;//sizeof(genes)/sizeof(genes[0]) ;
    //std::cout << "boom :"<<genes_size << "-->"<< random()<<std::endl;
    for (int i = 0; i < genes_size; i++) {
      if (random() < mutationRate) {
        genes[i] = random(32,127);
        //std::cout << "boom boom boom :"<<genes[i]<<std::endl;
        //std::cout << "bbmm:"<<genes[i]<<std::endl;
      }
    }
  }

char DNA::random(int min, int max)
{
    std::random_device rd; // obtain a random number from hardware
    std::mt19937 generator(rd()); // seed the generator
    std::uniform_int_distribution<> distr(min, max); // define the range

    return (char) distr(generator);
}


int DNA::random(int max)
{
    std::random_device rd; // obtain a random number from hardware
    std::mt19937 generator(rd()); // seed the generator
    std::uniform_int_distribution<> distr(0, max); // define the range

    return distr(generator);
}

float DNA::random()
{
    std::random_device rd; // obtain a random number from hardware
    std::mt19937 generator(rd()); // seed the generator
    std::uniform_real_distribution<> distr(0, 1); // define the range

    return (float) distr(generator);
}



std::string DNA::getPhrase() {

    //float* angles = new float[3];
    std::string _genes ;

 
    for (int i =0; i < gene_size; i++){
        _genes.push_back(genes[i]);

    }
    return _genes ;
}

